### PR TITLE
feat: Add VJP cache + `jax-cache` recipe

### PR DESCRIPTION
**Pull request by dionhaefner**
#### Relevant issue or PR

#455 
https://github.com/pasteurlabs/tesseract-jax/issues/27

#### Description of changes

Experimental cache to side-step the fact that we can't save residuals of the forward pass when computing VJPs naively.

#### Testing done

CI, manually confirmed increased throughput on a demo application

### CODE REVIEW

File: tesseract_core/sdk/templates/jax-cache/tesseract_api.py [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+# Jax-handled gradient endpoints (no need to modify)
+#
+
+
+def jacobian(
+    inputs: InputSchema,
+    jac_inputs: set[str],
+    jac_outputs: set[str],
+):
+    return jac_jit(inputs.model_dump(), tuple(jac_inputs), tuple(jac_outputs))
```
Comment (jpbrodrick89) [inline tesseract_core/sdk/templates/jax-cache/tesseract_api.py:123]: We can vmap over the cached vjp_func here too.

File: tesseract_core/sdk/templates/jax-cache/tesseract_api.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+    # to only differentiate through the array outputs.
+    def _apply_for_vjp(inputs_dict):
+        out = apply_jit(inputs_dict)
+        diff_out, static_out = eqx.partition(out, eqx.is_array)
+        return diff_out, static_out
+
+    diff_primals, vjp_func, static_primals = jax.vjp(
+        _apply_for_vjp, inputs_dict, has_aux=True
+    )
+    out = eqx.combine(diff_primals, static_primals)
```
Comment (jpbrodrick89) [inline tesseract_core/sdk/templates/jax-cache/tesseract_api.py:102]: I think we should actually just use `jax.linearize` and then we can cache both the forward and the backward pass (except for the linear transpose which I think is fairly cheap).

Comment (jpbrodrick89) [reply tesseract_core/sdk/templates/jax-cache/tesseract_api.py:102]: happy to take this on if you like

Comment (jpbrodrick89) [reply tesseract_core/sdk/templates/jax-cache/tesseract_api.py:102]: another advantage is that if we linearize and calculate the transpose later then we can safely rely on filter_func instead of caching the cotangent struct too.

Comment (dionhaefner) [reply tesseract_core/sdk/templates/jax-cache/tesseract_api.py:102]: Sure, please give it a shot

**Comment by jpbrodrick89** [comment]
Work continuing on #577
